mewarpx/mewarpx/emission.py

```diff
@@ -2049,3 +2065,156 @@ def _get_x_coords(self, npart):
+        interp_spec_1_dgrid = self._interpolate_grid(spec_1_dgrid)
+        self.spec_1_dgrid = (interp_spec_1_dgrid * mwxrun.dx * mwxrun.dz
+                             * (mwxrun.ymax - mwxrun.ymin)
+                             * 1e6).ravel()
+
+        # do same calculations for second grid if exists
+        if spec_2_dgrid is not None:
+            interp_spec_2_dgrid = self._interpolate_grid(spec_2_dgrid)
+            self.spec_2_dgrid = (interp_spec_2_dgrid * mwxrun.dx * mwxrun.dz
+                                 * (mwxrun.ymax - mwxrun.ymin)
+                                 * 1e6).ravel()
+        else:
+            self.spec_2_dgrid = None
+
+    @staticmethod
+    def _interpolate_grid(input_grid):
+        """Takes a grid (2d array) as an input and interpolates the values on
+        to the current simulation grid.
+
+        Arguments:
+            input_grid (np.ndarray): 2d array of any shape representing a
+                2d grid of plasma densities
+
+        Returns:
+            output_grid (np.ndarray): 2d array of current simulation dimensions
+                interpolated values from input_grid
+        """
+
+        input_shape = input_grid.shape
+        input_x = np.linspace(0.5, input_shape[0] - 0.5, input_shape[0])
+        input_z = np.linspace(0.5, input_shape[1] - 0.5, input_shape[1])
+        input_zz, input_xx = np.meshgrid(input_z, input_x)
+        input_coord = np.column_stack([input_xx.flatten(), input_zz.flatten()])
+
+        output_xx = input_xx * mwxrun.nx/input_shape[0]
+        output_zz = input_zz * mwxrun.nz/input_shape[1]
+
+        output_grid = scipy.interpolate.griddata(
+            input_coord, input_grid.flatten(), (output_xx, output_zz)
+        )
+        return output_grid
+
+    def _get_x_coords(self, npart):
+        """Uniformly samples particles in each grid cell, so each cell has the
+        same number of particles."""
+        nppc = npart / (mwxrun.nx * mwxrun.nz)
+
+        if not nppc.is_integer() or nppc < 1:
+            raise ValueError(f"When using ArbitraryDistributionVolumeEmitter, "
+                             f"number of particles per cell per species must "
+                             f"be an integer greater than or equal to 1. "
+                             f"Currently it is {nppc}.")
+
+        random_dx = np.random.uniform(-mwxrun.dx / 2, mwxrun.dx / 2, npart)
+        random_dz = np.random.uniform(-mwxrun.dz / 2, mwxrun.dz / 2, npart)
+
+        x = np.linspace(self.bounds[0, 0] + mwxrun.dx/2, self.bounds[0, 1] - mwxrun.dx/2, mwxrun.nx)
+        z = np.linspace(self.bounds[2, 0] + mwxrun.dz/2, self.bounds[2, 1] - mwxrun.dz/2, mwxrun.nz)
+        zz, xx = np.meshgrid(z, x)
+        xx = np.tile(xx.ravel(), int(nppc)) + random_dx
+        zz = np.tile(zz.ravel(), int(nppc)) + random_dz
+        yy = np.zeros_like(xx)
+
+        return np.array([xx, yy, zz]).T
+
+    def _gen_particle_dict(self, x, y, z, vx, vy, vz, w, species, **kwargs):
```

Comment:
We should be able to recode this as a `wfn` that's already called by default `_gen_particle_dict` and simplify this substantially.